### PR TITLE
mgr/dashboard: progress: support rbd_support module async tasks

### PR DESCRIPTION
**Pull request by rjfd**
This PR adds the support for rbd_support module async tasks by showing those tasks events (from the progress module) as dashboard rbd tasks.

In practice, when the user deletes an RBD image from the CLI using the async task from the rbd_support module, the dashboard frontend automatically shows that operation and it's progress in the rbd list page.

Signed-off-by: Ricardo Dias <rdias@suse.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard/services/progress.py [hunk 1]
```diff
@@ -14,37 +14,67 @@
...
+    if event['refs'] and isinstance(event['refs'], dict):
+        refs = event['refs']
+        if refs['origin'] == "rbd_support":
+            # rbd mgr module event, we can transform this event into an rbd dashboard task
+            action_map = {
+                'remove': "delete",
+                'flatten': "flatten",
+                'trash remove': "trash/remove"
+            }
+            action = action_map[refs['action']] if refs['action'] in action_map else refs['action']
```
Comment (dillaman) [inline src/pybind/mgr/dashboard/services/progress.py]: Nit: ```action = action_map.get(refs['action'], refs['action'])```?

File: src/pybind/mgr/dashboard/services/progress.py [hunk 2]
```diff
@@ -14,37 +14,67 @@
 from .. import mgr, logger
 
 
+def _progress_event_to_dashboard_task_common(event, task):
+    if event['refs'] and isinstance(event['refs'], dict):
```
Comment (dillaman) [inline src/pybind/mgr/dashboard/services/progress.py]: Nit: I don't think `event['refs']` is necessary. If `refs` isn't in the dict, you'd get a `KeyError` and if it's not a dict, it is being checked by the second condition.

**Comment by rjfd** [comment]
@dillaman I have two questions about the rbd_support tasks:

* why are the tasks schedule to be run sequentially? i.e., why can't we run a limited number of tasks in parallel?

* the current implementation checks if a task was already scheduled not only by looking at the ones executing (or waiting to execute) but also by looking at the list of tasks that already finished. This creates a non-intuitive behavior that can be reproduced with the following steps:
  1. create rbd image "disk1"
  2. schedule delete task (rbd task add remove disk1), and wait for it to finish
  3. create rbd image "disk1"
  4. schedule delete task (rbd task add remove disk1), <- does not do anything because it thinks it is replaying step 2.

Can we change this behavior?

**Comment by dillaman** [comment]
>     * why are the tasks schedule to be run sequentially? i.e., why can't we run a limited number of tasks in parallel?

Because there is an immidiate need to get this stuff backported to Mimic+Nautilus to support `ceph-csi`. Parallel tasks could be done in the future but it wasn't needed for initial support and would just delay the work product. 
 
>     * the current implementation checks if a task was already scheduled not only by looking at the ones executing (or waiting to execute) but also by looking at the list of tasks that already finished. This creates a non-intuitive behavior that can be reproduced with the following steps:

It's a requirement for all `ceph` CLI commands to be idempotent. See the last commit in my PR. It could be tweaked to also store image ids in the refs to help avoid your scenario.

**Comment by rjfd** [comment]
@dillaman thanks!

**Comment by runsisi** [comment]
hi @dillaman , i am wondering what will happen if the current active mgr fails over to another mgr? will those aync tasks be restarted? thanks.

**Comment by dillaman** [comment]
> hi @dillaman , i am wondering what will happen if the current active mgr fails over to another mgr? will those aync tasks be restarted? thanks.

Indeed. The tasks are persisted in a new `rbd_task` object in each pool/namespace. At MGR start-up, these tasks are reloaded from the OSD and rescheduled.

**Comment by runsisi** [comment]
@dillaman it makes sense. thank you!

**Comment by rjfd** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/services/progress.py (hunk 2 above)
Comment (rjfd) [reply src/pybind/mgr/dashboard/services/progress.py]: I don't want to raise an exception if `refs` is not in the event dict. I just want to proceed gracefully.

File: src/pybind/mgr/dashboard/services/progress.py (hunk 1 above)
Comment (rjfd) [reply src/pybind/mgr/dashboard/services/progress.py]: fixed

**Comment by callithea** [comment]
QA run [WIP]: http://pulpito.ceph.com/laura-2019-08-30_19:24:37-rados:mgr-wip-lpaduano-testing-29424-29532-distro-basic-smithi/

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
QA run [WIP]:
http://pulpito.ceph.com/laura-2019-09-09_12:05:24-rados:mgr-wip-lpaduano-testing-29424-30095-09-09-19-distro-basic-smithi/

**Pull request merged**